### PR TITLE
CRUD items with minimalist flux

### PR DESCRIPTION
**Pull request by leplatrem**
This PR could serve as a tutorial (or base branch) to add «edit in place» for the list.

This PR could remain open, regularly rebased. And developers could comment or ask questions. This would allow to keep the master boilerplate minimalistic, while providing well known use-cases.

Thoughts ?

### CODE REVIEW

File: scripts/components/App.js [hunk 1]
```diff
@@ -3,34 +3,88 @@ import React from "react";
...
+    var newrecord = Object.assign({}, this.state.record, {[field]: event.target.value});
+    this.setState({record: newrecord});
   }
 
   render() {
     return (
       <form onSubmit={this.onFormSubmit.bind(this)}>
-        <input name="label" type="text" />
-        <button type="submit">Add</button>
+        <input ref="label"
```
Comment (leplatrem) [inline scripts/components/App.js]: ref not used

File: scripts/components/App.js [hunk 2]
```diff
@@ -3,34 +3,88 @@ import React from "react";
...
+  constructor(props) {
+    super(props);
+    this.state = {record: this.props.record || {}};
+  }
+
   onFormSubmit(event) {
     event.preventDefault();
-    var record = {label: event.target.elements.label.value};
-    this.props.updateRecord(record);
+    var action = this.props.record ? 'update' : 'create';
```
Comment (n1k0) [inline scripts/components/App.js]: Nit: `const` (here and everywhere else applicable)

File: scripts/components/App.js [hunk 3]
```diff
@@ -3,34 +3,88 @@ import React from "react";
...
+    this.props.events.emit('action:' + action, this.state.record);
```
Comment (n1k0) [inline scripts/components/App.js]: Nit: `action:${action}`

File: scripts/components/App.js [hunk 4]
```diff
@@ -3,34 +3,88 @@ import React from "react";
...
+    super(props);
+    this.state = {current: null};
+  }
 
-  static get defaultProps() {
-    return {items: []};
+  componentDidMount() {
+    this.props.events.on("action:edit", (index) => {
+      this.setState({current: index});
+    });
```
Comment (n1k0) [inline scripts/components/App.js]: Nit:

``` js
this.props.events.on("action:edit", current => this.setState({current: current}));
this.props.events.on("action:update", _ => this.setState({current: null}));
this.props.events.on("action:delete", _ => this.setState({current: null}));
```

I wish we could subscribe to multiple events at once as with Backbone.Events… Oh well.

File: scripts/components/App.js [hunk 5]
```diff
@@ -42,32 +96,33 @@ export default class App extends React.Component {
...
     super(props);
-    this.state = this.props.store.state;
-    this.props.store.on("change", state => {
-      this.setState(Object.assign({busy: false}, state));
-    });
-    this.props.store.on("error", error => {
-      this.setState({busy: false, error: error.message});
-    });
-    this.props.store.load();
+    this.state = {busy: false, error: "", items: []};
```
Comment (n1k0) [inline scripts/components/App.js]: Please always use `setState`, see [important note here](https://facebook.github.io/react/docs/component-api.html#setstate).

File: scripts/components/App.js (hunk 4 above)
Comment (n1k0) [inline scripts/components/App.js]: Even better:

``` js
const updateCurrent = current => this.setState({current});
this.props.events.on("action:edit", updateCurrent);
this.props.events.on("action:update", updateCurrent);
this.props.events.on("action:delete", updateCurrent);
```

File: scripts/components/App.js [hunk 6]
```diff
@@ -42,32 +96,33 @@ export default class App extends React.Component {
...
 
   render() {
     var disabled = this.state.busy ? "disabled" : "";
     return (
       <div className={disabled}>
-        <Form updateRecord={this.updateRecord.bind(this)}/>
-        <List items={this.state.items.map(item => item.label)}/>
-        <button onClick={this.syncRecords.bind(this)} disabled={disabled}>Sync!</button>
+        <Form events={this.props.events}/>
+        <List events={this.props.events}
```
Comment (n1k0) [inline scripts/components/App.js]: We should really investigate passing this `events` prop transparently using a context :/

File: scripts/components/App.js [hunk 7]
```diff
@@ -42,32 +96,33 @@ export default class App extends React.Component {
...
+    });
+    this.props.events.on("store:error", error => {
+      this.setState({error: error.message});
+    });
   }
 
-  syncRecords() {
-    this.setState({busy: true, error: ""});
-    this.props.store.sync();
+  onSync() {
```
Comment (n1k0) [inline scripts/components/App.js]: Nit: `onSyncClick`; it's usually useful to have the event type part of the handling method name.

File: scripts/controller.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    this.events = events;
+
+    this.events.on('action:create', this.onCreate.bind(this));
+    this.events.on('action:update', this.onUpdate.bind(this));
+    this.events.on('action:delete', this.onDelete.bind(this));
+    this.events.on('action:sync', this.onSync.bind(this));
+  }
+
+  start() {
+    console.log('Start');
```
Comment (n1k0) [inline scripts/controller.js]: Nit: :)

File: scripts/controller.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+export class Controller {
+
+  constructor(stores, events) {
+    this.dataStore = stores.store;
+    this.events = events;
+
+    this.events.on('action:create', this.onCreate.bind(this));
+    this.events.on('action:update', this.onUpdate.bind(this));
+    this.events.on('action:delete', this.onDelete.bind(this));
+    this.events.on('action:sync', this.onSync.bind(this));
```
Comment (n1k0) [inline scripts/controller.js]: I feel like we could map event names to `dataStore` method calls, though that'd be probably a little too much implicit/magic…

File: scripts/store.js [hunk 10]
```diff
@@ -1,49 +1,72 @@
...
+    this.events.emit('store:busy', true);
+    return this.collection.delete(record.id)
+      .then(res => {
+        this.state.items = this.state.items.filter(item => {
+          return item.id !== record.id;
+        });
+      })
+      .then(this.onReady.bind(this))
+      .catch(this.onError.bind(this));
+  }
```
Comment (n1k0) [inline scripts/store.js]: I feel like the three crud methods could be factored a little further…

File: test/components/app_test.js [hunk 11]
```diff
@@ -119,9 +284,45 @@ describe("Form", () => {
...
+
+  describe("Editing", () => {
+
+    var callback;
+    const record = {id: "42", label: "Value"};
+
+    beforeEach(() => {
+      callback = sinon.spy();
+      rendered = TestUtils.renderIntoDocument(<Form record={record}
+                                                    saveRecord={callback} />);
```
Comment (n1k0) [inline test/components/app_test.js]: Nit: strange formatting; how about 

``` js
  rendered = TestUtils.renderIntoDocument(
    <Form record={record} saveRecord={callback} />);
```

File: test/store_test.js [hunk 12]
```diff
@@ -61,6 +61,34 @@ describe("Store", () => {
   });
 
 
+  describe("#update()", () => {
+
+    it("update() replaces with Kinto record and emits change", (done) => {
```
Comment (n1k0) [inline test/store_test.js]: Nit: uneecessary `update()` mention as we're already in a section mentioning it.

File: scripts/components/App.js (hunk 5 above)
Comment (leplatrem) [inline scripts/components/App.js]: In the constructor, I can assign the state !

Look https://facebook.github.io/react/blog/2015/01/27/react-v0.13.0-beta-1.html#es6-classes

**Comment by leplatrem** [comment]
- [x] Async auth
- [x] Use flux pattern
- [x] Configure components from events (_to be reviewed_)
- [ ] Rename `flux` to `dispatcher` ?

r? @n1k0

File: scripts/flux.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    names.forEach(event => {this.events.on(event, callback)});
+  }
+
+  dispatch(name, data) {
+    this.events.emit(name, data);
+  }
+
+  onConfigure(config) {
+    this.auth.configure(config);
+    this.auth.authenticate();
```
Comment (leplatrem) [inline scripts/flux.js]: Should this be : 
- a) `this.auth.authenticate(config)`
- b) `this.auth.configure(config).then(this.auth.authenticate)`
- c) like now
- d) `events.on('configure')` in auth class
  ?

File: scripts/flux.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
+export default class Flux {
```
Comment (leplatrem) [inline scripts/flux.js]: Rename to Dispatcher ?

File: scripts/index.js [hunk 15]
```diff
@@ -1,25 +1,21 @@
...
+const events = new EventEmitter();
+const auth = new Auth(events);
+const store = new Store("items", events);
+const flux = new Flux({auth, store}, events);
 
+React.render(<App flux={flux}/>, document.getElementById("app"));
 
-React.render(<App store={store}/>, document.getElementById("app"));
+// Start with default server.
+events.emit('action:configure', {server: DEFAULT_SERVER});
```
Comment (leplatrem) [inline scripts/index.js]: Should we replace this by a  `action:start` from main `App::componentDidMount()` ?

File: scripts/flux.js (hunk 14 above)
Comment (n1k0) [inline scripts/flux.js]: Not exactly, if we map the Flux concept here, the dispatcher role is actually played by the event emitter; Flux is the whole container for the stores and the dispatcher. But I totally get what you mean… Actually if we don't want too much to stick to the Flux lexical scope, _Controller_ would work pretty well too.

File: scripts/index.js (hunk 15 above)
Comment (n1k0) [inline scripts/index.js]: That's a great idea. You probably want to ensure the app starts when the UI is ready anyway.

File: scripts/flux.js (hunk 13 above)
Comment (n1k0) [inline scripts/flux.js]: Subscribing to an event seems more consistent with the rest of the code, I'd vote for that.

File: scripts/flux.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+  constructor(components, events) {
+    this.store = components.store;
+    this.auth = components.auth;
```
Comment (n1k0) [inline scripts/flux.js]: Nit: This may work (untested): 

``` js
{this.store, this.auth} = components;
```

File: scripts/flux.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    this.events.on('action:configure', this.onConfigure.bind(this));
+    this.events.on('action:create', this.onCreate.bind(this));
+    this.events.on('action:update', this.onUpdate.bind(this));
+    this.events.on('action:delete', this.onDelete.bind(this));
+    this.events.on('action:sync', this.onSync.bind(this));
+  }
+
+  on(events, callback) {
+    const names = events.split(/\s*,\s*/);
+    names.forEach(event => {this.events.on(event, callback)});
```
Comment (n1k0) [inline scripts/flux.js]: Nit: you don't need curly brackets for single statement lambdas:

``` js
names.forEach(event => this.events.on(event, callback)); 
```

File: scripts/flux.js (hunk 16 above)
Comment (leplatrem) [inline scripts/flux.js]:
```
SyntaxError: ..scripts/flux.js: Unexpected token (7:28)
   5 | 
   6 |   constructor(components, events) {
>  7 |     {this.store, this.auth} = components;
     |                             ^
   8 |     this.events = events;
```

Comment (n1k0) [inline scripts/flux.js]: Dommage :D

**Comment by leplatrem** [comment]
Visible online at http://kinto.github.io/kinto-react-boilerplate/

**Comment by n1k0** [comment]
> Visible online at http://kinto.github.io/kinto-react-boilerplate/

Side note, got this in the console:

_Warning: Each child in an array or iterator should have a unique "key" prop. Check the render method of t. See https://fb.me/react-warning-keys for more information._

File: scripts/auth.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    this.server = '';
+  }
+
+  configure(config) {
+    this.server = config.server;
+  }
+
+  authenticate() {
+    // Take username from location hash (With default value):
+    const user = window.location.hash = (window.location.hash.slice(1) || "public-demo");
```
Comment (n1k0) [inline scripts/auth.js]: `window` as an internal dependency makes this very hard to unit test if we want to; how about supporting injecting it through a constructor option?

File: scripts/components/App.js [hunk 19]
```diff
@@ -1,36 +1,97 @@
...
+    });
   }
 
   render() {
     return (
       <ul>{
-        this.props.items.map((label, i) => {
-          return <li key={i}>{label}</li>;
+        this.props.items.map((item, i) => {
+          return <Item index={i}
```
Comment (n1k0) [inline scripts/components/App.js]: A `key` prop is probably missing here, hence the warning we get about that.

**Comment by n1k0** [comment]
The patch looks great. I was surprised not seeing any kind of visual notification/feedback when the _sync_ operation is performed, though that may not be the purpose of this PR.

r=r+wc

**Comment by leplatrem** [comment]
> not seeing any kind of visual notification

Good catch !

I am not sure about I did for the `window.location`. Please don't give up :)

**Comment by n1k0** [comment]
> I am not sure about I did for the `window.location`. Please don't give up :)

Exactly what I had in mind :+1:

**Pull request merged**